### PR TITLE
Tickets/dm 3566

### PR DESCRIPTION
**Pull request by n8pease**
Adds the butler datasetType feature and supporting tests.
If adding the data files to this project is not the best approach let's discuss other ways to do it.

### CODE REVIEW

File: python/lsst/daf/persistence/butler.py [hunk 1]
```diff
@@ -136,8 +138,15 @@ def __init__(self, root, mapper=None, **mapperArgs):
         self.log = pexLog.Log(pexLog.Log.getDefaultLog(),
                 "daf.persistence.butler")
 
-    def getKeys(self, datasetType=None, level=None):
+    def registerDatasetTypeAlias(self, type, alias):
```
Comment (ktlim) [inline python/lsst/daf/persistence/butler.py]: Use of `type` is discouraged as it is a Python built-in.  I tend to use `datasetType` instead.

File: python/lsst/daf/persistence/butler.py [hunk 2]
```diff
@@ -433,8 +451,18 @@ def _read(self, pythonType, location):
         return results
 
     def __reduce__(self):
-        return (_unreduce, (self.mapper,))
+        return (_unreduce, (self.mapper, self.datasetTypeAliasDict))
+
+    def __resolveDatasetTypeAlias(self, type):
```
Comment (ktlim) [inline python/lsst/daf/persistence/butler.py]: You should generally use a single underscore to indicate a "private/protected" method, not a double underscore (which, among other things, prevents overriding in a subclass).

File: python/lsst/daf/persistence/butler.py [hunk 3]
```diff
@@ -433,8 +451,18 @@ def _read(self, pythonType, location):
...
+        resolvedType = type
```
Comment (ktlim) [inline python/lsst/daf/persistence/butler.py]: I think this whole method could be replaced by `return self.datasetTypeAliasDict.get(datasetType, datasetType)`.  But it doesn't quite do enough.  First of all, the specification is to replace "`@alias`" (not just "`alias`") with "`datasetType`".  Second, the specification is also to replace "`@alias`" when it occurs as a *substring* of the original type: "`@coadd`" -> "`deepStack`" also implies "`@coadd_src`" -> "`deepStack_src`".

**Comment by n8pease** [comment]
Fritz messed up this PR -- starting a new one

**Pull request closed without merging**